code/modules/unit_tests/blackmarket.dm

```diff
@@ -0,0 +1,21 @@
+/// Ensures black market items have acceptable variable values.
+/datum/unit_test/blackmarket
+
+/datum/unit_test/blackmarket/Run()
+	for(var/datum/market_item/prototype as anything in subtypesof(/datum/market_item))
+		if(initial(prototype.abstract_path) == prototype) //skip abstract paths
+			continue
+		if(!initial(prototype.category))
+			TEST_FAIL("[prototype] doesn't have a set category (or the abstract path var isn't correctly set)")
+		if(!initial(prototype.item))
+			TEST_FAIL("[prototype] doesn't have a set item (or the abstract path var isn't correctly set)")
+		if(isnull(initial(prototype.price)) && initial(prototype.price_max) <= initial(prototype.price_min))
+			TEST_FAIL("[prototype] doesn't have a correctly set random price (price_max should be higher than price_min)")
+		if(isnull(initial(prototype.stock)) && initial(prototype.stock_max) < initial(prototype.stock_min))
+			TEST_FAIL("[prototype] doesn't have a correctly set random stock (stock_max shouldn't be lower than stock_min)")
+		if(!isnum(initial(prototype.availability_prob)))
+			TEST_FAIL("[prototype] doesn't have a set availability_prob (must be a number)")
+		if(!initial(prototype.name))
+			TEST_FAIL("[prototype] doesn't have a set name")
+		if(!initial(prototype.desc))
+			TEST_FAIL("[prototype] doesn't have a set desc")
```

Comment:
We should be using ASSERTS in unit tests. also the fingerprint is slightly cleaner with the `::` operator

```suggestion
		if(prototype::abstract_path == prototype) //skip abstract paths
			continue
		TEST_ASSERT_NOTNULL(prototype::name, "[prototype] doesn't have a set name")
		TEST_ASSERT_NOTNULL(prototype::desc, "[prototype] doesn't have a set desc")
		TEST_ASSERT_NOTNULL(prototype::category, "[prototype] doesn't have a set category (or the abstract path var isn't correctly set)")
		TEST_ASSERT_NOTNULL(prototype::item, "[prototype] doesn't have a set item (or the abstract path var isn't correctly set)")
		TEST_ASSERT(isnum(prototype::availability_prob), "[prototype] doesn't have a set availability_prob (must be a number)")
		TEST_ASSERT(prototype::price && (prototype::price_max >= prototype::price_min), "[prototype] doesn't have a correctly set random price (price_max should be higher than price_min)")
		TEST_ASSERT(prototype::stock&& (prototype::stock_max > prototype::stock_min), "[prototype] doesn't have a correctly set random stock (stock_max shouldn't be lower than stock_min)")	
```

Reply:
the code should work but we'll know if it doesn't

Reply:
We cannot use the ASSERT macros here, since they call `return` when they don't pass, thus stopping the loop early.
I'll go and apply the `::` operator though.

Reply:
lol that seems like an oversight. maybe we could do something like
```dm
/datum/unit_test/thing/Run()
 for(var/type/x as anything in subtypesof(whatever))
  if(type == abstract_type)
   continue
  validate_datum(type)

/// runs test assertions on a prototype
/datum/unit_test/thing/proc/validate_datum(prototype)
 ...
```